crates/matrix-sdk-crypto/src/secret_storage.rs

```diff
@@ -245,29 +245,39 @@ impl SecretStorageKey {
     fn check_zero_message(&self) -> Result<(), DecodeError> {
         match &self.storage_key_info.algorithm {
             SecretStorageEncryptionAlgorithm::V1AesHmacSha2(properties) => {
-                if properties.iv.as_bytes().len() != IV_SIZE {
-                    Err(DecodeError::IvLength(IV_SIZE, properties.iv.as_bytes().len()))
-                } else {
-                    let mut iv_array = [0u8; 16];
-                    iv_array.copy_from_slice(properties.iv.as_bytes());
-
-                    // I'm not particularly convinced that this couldn't have been done simpler. Why
-                    // do we need to reproduce the ciphertext? Couldn't we just generate the MAC tag
-                    // using the `ZERO_MESSAGE`?
-                    //
-                    // If someone is reading this and is designing a new secret encryption
-                    // algorithm, please consider the above suggestion.
-                    let key = AesHmacSha2Key::from_secret_storage_key(&self.secret_key, "");
-                    let ciphertext = key.apply_keystream(Self::ZERO_MESSAGE.to_vec(), &iv_array);
-                    let expected_mac = HmacSha256Mac::from_slice(properties.mac.as_bytes())
-                        .ok_or_else(|| {
-                            DecodeError::MacLength(MAC_SIZE, properties.mac.as_bytes().len())
-                        })?;
-
-                    key.verify_mac(&ciphertext, expected_mac.as_bytes())?;
-
-                    Ok(())
+                let Some(iv) = &properties.iv else {
+                    return Err(DecodeError::IvLength(IV_SIZE, 0));
+                };
+
+                let Some(mac) = &properties.mac else {
+                    return Err(DecodeError::MacLength(MAC_SIZE, 0));
+                };
```

Comment:
For the record, I think the correct way to handle this is something like:

```suggestion
                let Some(iv), Some(mac) = (&properties.iv, &properties.mac) else {
                    // The IV and/or MAC are missing from the account data content. As the [spec] says, we have
                    // to assume that the key is valid.
                    //
                    // [spec]: https://spec.matrix.org/unstable/client-server-api/#msecret_storagev1aes-hmac-sha2
                    return Ok(());
                };
```

However, if you want to land this as-is, I'll not object.

Reply:
I've included your suggestion, thanks! The correct syntax is `let (Some(iv), Some(mac)) = …` though, otherwise I would have accepted it directly from the Github UI.

Waiting for the CI to validate this change before merging.